heat/core/arithmetics.py

```diff
@@ -197,6 +203,66 @@ def bitwise_xor(t1, t2):
     return operations.__binary_op(torch.Tensor.__xor__, t1, t2)
 
 
+def cumprod(a, axis):
+    if axis is None:
+        raise NotImplementedError("axis = None is not supported")
+    if not isinstance(axis, int):
+        raise RuntimeError("axis parameter got an invalid argument {}".format(axis))
+
+    cprod = torch.cumprod(a._DNDarray__array, axis=axis)
+
+    if a.split is not None and axis == a.split:
+        send = torch.ones(cprod.shape[:axis] + cprod.shape[axis + 1 :], dtype=cprod.dtype)
+
+        if cprod.numel() > 0:
+            indices = -1
+            Ni, Nk = cprod.shape[:axis], cprod.shape[axis + 1 :]
+            for ii in np.ndindex(Ni):
+                for kk in np.ndindex(Nk):
+                    send[ii + np.s_[...,] + kk] = cprod[ii + np.s_[:,] + kk][indices]
+
+        recv = torch.ones(cprod.shape[:axis] + cprod.shape[axis + 1 :], dtype=send.dtype)
+
+        a.comm.Exscan(send, recv, MPI.PROD)
```

Comment:
@krajsek does this work with the upcoming AD stuff?

Reply:
Not yet.